platform-sdk/docs/proposals/TSS-Encryption-Key-Automation/TSS-Encryption-Key-Automation.md

```diff
@@ -0,0 +1,233 @@
+  * This new transaction is system generated only and not user generated.  It should be rejected at the HAPI gateway
+    for user generated transactions.
+
+#### Alternatives Considered
+
+The alternative to the automated lifecycle is to have the node operators manually generate the `tssEncryptionKey`
+before the node starts and provide the public key to the network through a transaction signed by the node admin key.
+
+Requiring human intervention for this key increases friction that can easily be avoided. This saves all the testing
+labor related to testing human workflows.
+
+---
+
+## Changes
+
+The following changes to the system are required.
+
+### Architecture and/or Components
+
+The `TssBaseService` already exists with its own state.
+
+* This proposal adds a new `Map<Long, Bytes>` to the state for storing the public key of each node.
+* This proposal modifies the normal execution of the `TssBaseService` when its current public `tssEncryptionKey` is not
+  in the state.
+
+Cryptography Changes:
+
+* The `KeysAndCerts.java` record is extended with `byte[] privateTssEncryptionKey` and `byte[] publicTssEncryptionKey`.
+* The cryptography system adds a dependency on the TSS-Library to generate `tssEncryptionKey` pairs.
+
+### Core Behaviors
+
+There are two new behaviors that will be added to the system.
+
+1. At `System Startup`, during cryptography loading, the `tssEncryptionKey` will be loaded from disk or generated if it
+   does not exist.
+2. At the start of `TssBaseService` execution, the `TssBaseService` will verify that the public `tssEncryptionKey`
+   is in the state, or possibly generate a new `tssEncryptionKey` pair and send a transaction to update the public
+   key in the state.
+
+#### Loading Cryptography
+
+The cryptography library will add a dependency on the `Tss-Library` to generate new public/private `tssEncryptionKeys`.
+
+In the `EnhancedKeystoreLoader`:
+
+1. During the `scan()` phase, the private `tssEncryptionKey` will be loaded from disk if present.
+2. During the `generate()` phase, if the private `tssEncryptionKey` is not present:
+   1. Create a new public/private key pair.
+   2. Write the private key to disk.
+
+At the end of loading cryptography, the `KeysAndCerts` record will be in one of two states:
+
+1. The private `tssEncryptionKey` is loaded from disk and there is no public `tssEncryptionKey`.
+2. The private and public `tssEncryptionKey` were generated, are present, and match each other.
+
+#### TssBaseService Initialization
+
+At the initialization of the `TssBaseService` the loaded public/private `tssEncryptionKey` will be provided. The
+private key will always be provided, but the public key may be null. There are three scenarios:
+
+1. The public `tssEncryptionKey` is provided indicating a new key-pair was generated.
+   1. Once execution is live, a `TssEncryptionKeyTransaction` will need to be sent periodically until the state
+      reflects the correct public key.
```

Comment:
Does the "sent periodically" part essentially implement a retry mechanism? Should there be a limit on the number of retries to avoid flooding the network with these transactions?

If it's not a retry mechanism, then why does this have to be done multiple times? Why a single transaction not sufficient?

Reply:
retry until the public key is committed to state.   

There are edge cases where events might become ancient before they come to consensus and their transactions are not processed. 

Avoid flooding by making the timeout period large, like 5 minutes.